Add NEON-SVE Bridge (#72)
ARM-software/acle: main/acle.rst

* Introduce ``__ARM_FEATURE_PAUTH`` and ``__ARM_FEATURE_BTI`` in sections
  ssec-PAC_ and ssec-BTI_ respectively.

Changes for next-release
~~~~~~~~~~~~~~~~~~~~~~~~

> * Add specification for NEON-SVE Bridge.

Reviewer: Please link the text of NEON-SVE Bridge to the `_ssec-NEONSVEBridge:` anchor.

ACLE support for SVE is defined in the Arm C Language Extensions for SVE
document [SVE-ACLE]_ available on the Arm Developer Website.

.. _ssec-NEONSVEBridge:

NEON-SVE Bridge
---------------

The NEON_SVE Bridge adds intrinsics that allow conversions between NEON and
SVE vectors.

> These intrinsics are defined in the header file ``arm_extensions_bridge.h``.

Reviewer 2: Perhaps this should be called something like arm_sve_neon.h or arm_neon_sve.h instead.  I think it will/should include both arm_neon.h and arm_sve.h.

The problem with a general header file like arm_extensions_bridge.h is that, over time, the header might accrue other unrelated features that the user wasn't expecting, possibly “infecting” the user's namespace with unforeseen styles of identifier.  For xample, suppose “arm_extensions_bridge.h” already existed for an earlier feature and that we then added the Neon/SVE bridge to it.  Users of the original version who upgrade their toolchain might find that arm_neon.h and arm_sve.h are now being included unexpectedly.

Reviewer 3: > Perhaps this should be called something like arm_sve_neon.h or arm_neon_sve.h instead. I think it will/should include both arm_neon.h and arm_sve.h.
If it does include those headers internally, will this not enlarge the total bytes processed by the compiler a fair amount? In total for clang this is 4MiB of source. If this header includes it, and the user includes it, then the compiler is going to have to pull in 8MiB of source. Even if this is not fully parsed it does represent a fair amount of extra work for every compilation unit which might include this header. I don't know for sure if this is worth being concerned about, but it seems we can make that problem go away if the new header forward declares the types it needs internally. Do you have any opinions on this problem and proposed solution?

Reviewer 2: I think it's fair to assume that people will only use the Neon/SVE bridge if they are also using Neon and SVE individually.  So from a user-interface perspective, I think we should document in the ACLE that including the new header file also has the effect of including arm_neon.h and arm_sve.h (just like including arm_sve.h is defined to have the effect of including arm_bf16.h when bfloat16 is supported).  Users can then write:

```
#include <new_header.h>
```

rather than having to write:

```
#include <arm_neon.h>
#include <arm_sve.h>
#include <new_header.h>
```

Even if they do (redundantly) include all three, I think the preprocessor has an optimisation that avoids reading a file a second time if all non-writespace/comment characters were protected by a macro guard.  (At least, GCC does this.  I assume Clang does too. :-))

Reviewer 3: 👍 thanks for the input, SGTM.

Author: I've changed the header file's name from `arm_extensions_bridge.h` to `arm_neon_sve_bridge.h`. I've updated the spec to say it now includes `arm_neon.h` and `arm_sve.h`

ACLE support for SVE is defined in the Arm C Language Extensions for SVE
document [SVE-ACLE]_ available on the Arm Developer Website.

.. _ssec-NEONSVEBridge:

NEON-SVE Bridge
---------------

The NEON-SVE Bridge adds intrinsics that enables setting NEON vectors
into SVE vectors, and getting NEON vectors from SVE vectors. As the first 128

> bit subvector of an SVE vector is physically a NEON register,

Reviewer 3: ```suggestion
bit subvector of an SVE vector physically overlaps with a NEON register,
```
('is' is a strong word, and the two register classes are distinct)

Author: I've removed this sentence entirely after an internal discussion

these intrinsics result in a no-op.

These intrinsics are defined in the header file ``arm_extensions_bridge.h``

> and must be included after ``arm_neon.h`` and ``arm_sve.h``. These

Reviewer 3: Good point. How about an approach along the lines of forward declaring the types used by the bridge so that the header stands alone? It may be a small thing but it doesn't seem ideal to `#include` the whole header multiple times into the same compilation unit if it is not necessary.

> and must be included after ``arm_neon.h`` and ``arm_sve.h``. These

Author: I was thinking more along the lines of
`#define __ARM_NEON_SVE_BRIDGE (__ARM_FEATURE_NEON && __ARM_FEATURE_SVE)``.
``__ARM_FEATURE_NEON` rather than including the header files, but yes the forced order isnt ideal

> and must be included after ``arm_neon.h`` and ``arm_sve.h``. These

Author: or something like
```
#if !defined(__ARM_FEATURE_NEON) || !defined(__ARM_FEATURE_SVE)
  error
#else
  #define __ARM_NEON_SVE_BRIDGE 1
```

> and must be included after ``arm_neon.h`` and ``arm_sve.h``. These

Reviewer 3: Forward-declaring the types sounds like the most palatable solution to me at the moment to make the headers work regardless of order (principle of least surprise for the user), and to avoid double-including them or anything like that.

> and must be included after ``arm_neon.h`` and ``arm_sve.h``. These

Reviewer: Resolving this conversation as I understand we have decided to force the inclusion of `arm_sve.h` and `arm_neon.h` directly in the header file of the bridge.

intrinsics are available in the ``aarch64`` state.

Including this header file will define ``__ARM_SVE_NEON_BRIDGE`` as 1.

``svset_neonq``
~~~~~~~~~~~~~~~

These intrinsics set NEON vector ``subvec`` into the first 128 bit subvector of

> the SVE vector ``vec``.

Reviewer 4: Is it worth including a version of these to set the low 128-bits of an SVE vector and ignore the top bits entirely. I imagine we can do that with this existing proposal with e.g. `svset_neonq_s8(svundef_s8(), a)` but I don't know whether there is a preference for relying on compilers optimising that away into a no-op versus providing a dedicated set of intrinsics to handle this?

Author: `svset_neonq_s8(svundef_s8(), a)` should be able to handle this perfectly fine we think.

|                                                                                |
|      svuint64_t svset_neonq[_u64](svuint64_t vec, uint64x2_t subvec)           |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svfloat16_t svset_neonq[_f16](svfloat16_t vec, float16x8_t subvec)        |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svfloat32_t svset_neonq[_f32](svfloat32_t vec, float32x4_t subvec)        |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svfloat64_t svset_neonq[_f64](svfloat64_t vec, float64x2_t subvec)        |

> +--------------------------------------------------------------------------------+

Reviewer 4: Can you also add instances of both get and set for the `poly8x16_t` , `poly16x8_t` , `poly64x2_t` and `poly128_t` types?

Author: Any suggestions for the `poly128_t` signature? There's no `svint128_t`

Reviewer 4: I didn't notice that the polynomial types don't exist anymore! I would guess probably the corresponding unsigned types should be used (to match the `pmull{b,t}` intrinsics), with `poly128_t` corresponding to `svuint64x2_t` based on the signature of e.g. `svpmullb_pair`? Would prefer a more authoritative opinion though, @fpetrogalli?

Reviewer 2: Personally, I think we should only provide functions for element types that exist in both Neon and SVE.  The user can reinterpret the Neon type where necessary.

Author: I've decided to not to include polynomial types for now

Reviewer: May I ask you to add this explanation in the design document?

```
Why polynomial type are not supported
---------------------

We only provide functions for element types that exist in both Neon and SVE.
The user can reinterpret the Neon type where necessary.

```

| .. code:: c                                                  |
|                                                              |
|      int8x16_t svget_neonq[_s8](svint8_t vec)                |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      int16x8_t svget_neonq[_s16](svint16_t vec)              |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      int32x4_t svget_neonq[_s32](svint32_t vec)              |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      int64x2_t svget_neonq[_s64](svint64_t vec)              |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |

> |      uint8x16_t svget_neonq[_s8](svuint8_t vec)              |

Reviewer 5: Typo here and below:  `s` should be `u`.

Author: Thanks :)

> |      uint8x16_t svget_neonq[_s8](svuint8_t vec)              |

Reviewer 5: ```suggestion
|      uint8x16_t svget_neonq[_u8](svuint8_t vec)              |
```

+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |

> |      uint16x8_t svget_neonq[_s16](svuint16_t vec)            |

Reviewer 5: ```suggestion
|      uint16x8_t svget_neonq[_u16](svuint16_t vec)            |
```

+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |

> |      uint32x4_t svget_neonq[_s32](svuint32_t vec)            |

Reviewer 5: ```suggestion
|      uint32x4_t svget_neonq[_u32](svuint32_t vec)            |
```

+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |

> |      uint64x2_t svget_neonq[_s64](svuint64_t vec)            |

Reviewer 5: ```suggestion
|      uint64x2_t svget_neonq[_u64](svuint64_t vec)            |
```

+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      float16x8_t svget_neonq[_f16](svfloat16_t vec)          |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      float32x4_t svget_neonq[_f32](svfloat32_t vec)          |
+--------------------------------------------------------------+
| .. code:: c                                                  |
|                                                              |
|      float64x2_t svget_neonq[_f64](svfloat64_t vec)          |
+--------------------------------------------------------------+

> ``svdup_neonq``

Reviewer: I think we need a section in the design document explaining why we are adding these.

@rsandifo-arm wrote:

> It might be worth having “svdup_neonq” functions to duplicate a neon vector to fill an SVE one. This is just another way of writing svdupq_lane, without having to cast from Neon to SVE as a separate step. I see you've considered and rejected the possibility of extracting a Neon vector from an arbitrary point in the vector, and I think that's fair. The reasons I think the dup functions are nevertheless worthwhile are that:
>    * without it, the user would have to explicitly choose how to initialise the upper bits of the SVE register (in the svset_neonq call) only to overwrite those bits immediately.
>    * providing the function is consistent with svdup for scalar floating-point values, which are likewise implemented using the lane form of DUP.

Please summarize this in a new subsection in the design document.

+---------------------------------+----------------+------------+-----------------+

These intrinsics are available when ``arm_acle.h`` is included.

> Architectural Bridges

Reviewer: nit: can we choose a common title for the top section on bridges? I vote for  `Architectural Extension Bridges` over `Architectural Bridges`.

```suggestion
Architectural Extension Bridges
```

+---------------------------------+----------------+------------+-----------------+

These intrinsics are available when ``arm_acle.h`` is included.

Architectural Extension Bridges
###############################

.. _ssec-NEONSVEBridge:

NEON-SVE Bridge
===============

The NEON_SVE Bridge adds intrinsics that allow conversions between NEON and
SVE vectors.

These intrinsics are defined in the header file ``arm_neon_sve_bridge.h``.

Including this header_file will include ``arm_neon.h`` and ``arm_sve.h``.

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Reviewer: Yep - good point @rsandifo-arm ! I think that we should raise an error if the header file is included but the compiler is not defining the `__ARM_NEON_SVE_BRIDGE`. The macro should be defined every time SVE is present, therefore it seems to me that it should have the same meaning of `__ARM_FEATURE_SVE`. 

This might mean that we don't need to use any feature macro, and make the bridge available unconditionally every time we include the header file and we target SVE or SVE2 with the compiler?

Given that the bridge will include `arm_sve.h`, the error will be pretty clear if we want to use the bridge in a compiler invocation that is not targeting SVE.

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Reviewer 4: I guess the question is how best to allow the user to know whether `#include "arm_neon_sve_bridge.h"` will work or not, given that compilers with `arm_sve.h` and without the bridge already exist. If `__has_include` or similar is present then I agree it is possibly redundant, but I don't know if there is a precedent for relying on that?

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Reviewer 2: Yeah, AIUI, the feature macros are provided for the reason that @georges-arm mentioned.  E.g. the canonical way of including `arm_sve.h` is:
```
#ifdef __ARM_FEATURE_SVE
#include <arm_sve.h>
#endif
```
I'm pretty sure no-one does that, but it's the canonical way nonetheless.

The idea would be the same here.  The compiler defines `__ARM_NEON_SVE_BRIDGE` to advertise to the source code that the compiler provides an `arm_neon_sve_bridge.h`.  The code can then include the header file when available and (if it wants to) fall back to something else otherwise.

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Reviewer: I see, OK. So we do the following:

1. we introduce the macro in the specs but not in the header file;
2. we require the compiler to set it to 1 when targeting SVE;
3. We raise an error if the macro is not active and the header file is included via `#ifndef __ARM_NEON_SVE_BRIDGE #error "..." #endif` in the header file.

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Reviewer 2: 1. 👍 
2. We're probably in agreement and I'm probably just being overly pedantic, but: compilers should only set the macro to 1 if they provide the new header file.  Old compilers that target SVE would still be conforming (they would not define the macro and would not provide the header file).
3. I don't think the error is necessary.  The macro is only there to tell the user whether a feature is available.

> Including this header file will define ``__ARM_NEON_SVE_BRIDGE`` as 1.

Author: done.

``svset_neonq``
---------------

> These intrinsics set NEON vector ``subvec`` into the first 128 bit subvector of

Reviewer 2: Summarising our off-line discussion:

I agree with @peterwaller-arm that if we're not sure, specifying that the functions copy bit-for-bit (as a .Q) rather than lane-for-lane seems safer.  Users will then have a direct way of getting the behaviour they want, whether that's no code or a shuffle (using an additional Neon intrinsic call).  If instead we define the intrinsics to include a shuffle then users who don't want one would have to do a compensating shuffle and rely on the compiler optimising the sequence away.

If it turns out that the shuffle version is more generally useful, we could add new bridge functions that do that in future.  They would behave the same as the current ones on little-endian.

> These intrinsics set NEON vector ``subvec`` into the first 128 bit subvector of

Author: Resolving this and https://github.com/ARM-software/acle/pull/72#discussion_r751218453 as resolved then.

> These intrinsics set NEON vector ``subvec`` into the first 128 bit subvector of

Author: Added a section to the design document detailing this decision

| .. code:: c                                                                    |
|                                                                                |
|      svuint16_t svset_neonq[_u16](svuint16_t vec, uint16x8_t subvec)           |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svuint32_t svset_neonq[_u32](svuint32_t vec, uint32x4_t subvec)           |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svuint64_t svset_neonq[_u64](svuint64_t vec, uint64x2_t subvec)           |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |

> |      svfloat16_t svset_neonq[_f16](svfloat16_t vec, float16x8_t subvec)        |

Reviewer 2: I think there should be a bfloat16 version too.  For SVE, this kind of _bf16 function is available unconditionally.

Author: done.

+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svfloat32_t svset_neonq[_f32](svfloat32_t vec, float32x4_t subvec)        |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svfloat64_t svset_neonq[_f64](svfloat64_t vec, float64x2_t subvec)        |
+--------------------------------------------------------------------------------+

``svget_neonq``
---------------

> These intrinsics return a NEON vector from the first 128 bit subvector of

Reviewer 2: Same comments here as for `vset_neonq`:

* we need to define what happens for big-endian targets
* there should be a bf16 version too

Author: added bf16 version. big-endian isnt yet resolved.

| .. code:: c                                                                    |
|                                                                                |
|      svuint16_t svdup_neonq[_u16](uint16x8_t vec)                              |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svuint32_t svdup_neonq[_u32](uint32x4_t vec)                              |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |
|      svuint64_t svdup_neonq[_u64](uint64x2_t vec)                              |
+--------------------------------------------------------------------------------+
| .. code:: c                                                                    |
|                                                                                |

> |      svfloat16_t svdup_neonq[_f16](float16x8_t vec)                            |

Reviewer 2: bf16 here too

Author: done.

+---------------------------------+----------------+------------+-----------------+

These intrinsics are available when ``arm_acle.h`` is included.

Architectural Extension Bridges
###############################

.. _ssec-NEONSVEBridge:

NEON-SVE Bridge
===============

The NEON_SVE Bridge adds intrinsics that allow conversions between NEON and
SVE vectors.

These intrinsics are defined in the header file ``arm_neon_sve_bridge.h``.

``__ARM_NEON_SVE_BRIDGE`` is defined as 1 when ``arm_neon_sve_bridge.h`` is
available.

Including this header_file will include ``arm_neon.h`` and ``arm_sve.h``.

``svset_neonq``
---------------

> These intrinsics set the first 128 bits of SVE vector ``vec`` to ``subvec``.

Reviewer 3: I've had a request to mention that passing `svundef` as the `vec` parameter allows the compiler to reuse the SVE register corresponding to the NEON register. Here is some suggested language, please can someone raise their hand if they can improve it :)

```suggestion
These intrinsics set the first 128 bits of SVE vector ``vec`` to ``subvec``.
When ``svundef`` is passed as the ``vec`` parameter, the compiler is able
reuse the SVE register overlapping the NEON input without generating
additional instructions.
```

Reviewer: Sorry, I don' understand what is the issue that this change of wording is trying to address.

Reviewer 3: My intent was to try and provide a hint to the user how they might use this. But maybe this is not an appropriate place to do that, so I am happy to drop this suggestion. I found it difficult to shoehorn it in here with appropriate language.

Reviewer 2: The change looks good to me FWIW.  Might be worth making it a separate paragraph though.

> These intrinsics set the first 128 bits of SVE vector ``vec`` to ``subvec``.

Reviewer 2: I think this is subtle enough that it would be worth expanding.

```suggestion
These intrinsics set the first 128 bits of SVE vector ``vec`` to ``subvec``.
That is, bit *i* of the result is equal to:
* bit *i* of ``subvec`` if *i* < 128
* bit *i* of ``vec`` otherwise

.. note::

   On big-endian targets, this leaves lanes in a different
   order from the “native” SVE order.  For example, if ``subvec`` is
   ``int32x4_t``, then on big-endian targets, the first memory element
   is in lane 3 of ``subvec`` and is therefore in lane 3 of the returned
   SVE vector.  Using ``svld1`` to load elements would instead put the
   first memory element in lane 0 of the returned SVE vector.
```
